general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
 #define MFEM_MEM_MANAGER_HPP
 
 #include "globals.hpp"
+#include "error.hpp"
+#include <cstring> // std::memcpy
+#include <type_traits> // std::is_const
 
 namespace mfem
 {
 
 // Implementation of MFEM's lightweight device/host memory manager designed to
 // work seamlessly with the OCCA, RAJA, and other kernels supported by MFEM.
 
+/// Memory types supported by MFEM.
+enum class MemoryType
+{
+   HOST,      ///< Host memory; using new[] and delete[]
+   HOST_32,   ///< Host memory alligned at 32 bytes
+   HOST_64,   ///< Host memory alligned at 64 bytes
+   CUDA,      ///< cudaMalloc, cudaFree
+   CUDA_UVM   ///< cudaMallocManaged, cudaFree
+};
+
+/// Memory classes identify subsets of memory types.
+/** This type is used by kernels that can work with multiple MemoryType%s to
+    request access to any input/output Memory objects. */
+enum class MemoryClass
+{
+   HOST,    ///< Memory types: { HOST, HOST_32, HOST_64, CUDA_UVM }
+   HOST_32, ///< Memory types: { HOST_32, HOST_64 }
+   HOST_64, ///< Memory types: { HOST_64 }
+   CUDA,    ///< Memory types: { CUDA, CUDA_UVM }
+   CUDA_UVM ///< Memory types: { CUDA_UVM }
+};
+
+/// Return true is the given memory type is in MemoryClass::HOST.
+inline bool IsHostMemory(MemoryType mt) { return mt <= MemoryType::HOST_64; }
+
+/// Return a suitable MemoryType for a given MemoryClass.
+MemoryType GetSuitableMemoryType(MemoryClass mc);
+
+/// Return a suitable MemoryClass from a pair of MemoryClass%es.
+/** Note: this operation is commutative, i.e. a*b = b*a, associative, i.e.
+    (a*b)*c = a*(b*c), and has an identity element: MemoryClass::HOST.
+
+    Currently, the operation is defined as a*b := max(a,b) where the max
+    operation is based on the enumeration ordering:
+
+        HOST < HOST_32 < HOST_64 < CUDA < CUDA_UVM. */
+MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
+
+/// Class used by MFEM to store pointers to host and/or device memory.
+/** The template class paramter, T, must be a plain-old-data (POD) type.
+
+    In many respects this class behaves like a pointer:
+    * When destroyed, a Memory object does NOT automatically delete any
+      allocated memory.
+    * Only the method Delete() will deallocate a Memory object.
```

Comment:
This is part of the design -- `Memory` (like pointers) has to be explicitly allocated and de-allocated.

Reply:
What I meant is why not the standard constructor/destructor? What is the benefit in keeping these objects instead of making new ones and deleting them?

Reply:
I finally got it.